CMakeLists unconditionally force -fobjc-arc flag for Apple which breaks build with GCC

### What is the issue?

```
cd /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/build/libtransmission && /opt/local/bin/g++-mp-12 -DDISABLE_GETTEXT -DFMT_EXCEPTIONS=0 -DFMT_HEADER_ONLY=1 -DHAVE_COPYFILE -DHAVE_FLOCK -DHAVE_MKDTEMP -DHAVE_PREAD -DHAVE_PWRITE -DHAVE_SO_REUSEPORT=1 -DHAVE_STATVFS -DHAVE_SYS_STATVFS_H -DHAVE_XLOCALE_H -DNATPMP_STATICLIB -DPACKAGE_DATA_DIR=\"/opt/local/share\" -DPOSIX -DSYSTEM_MINIUPNP -DUSE_SYSTEM_B64 -DWIDE_INTEGER_DISABLE_FLOAT_INTEROP -DWIDE_INTEGER_DISABLE_IOSTREAM -DWITH_KQUEUE -DWITH_UTP -D__TRANSMISSION__ -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/build/libtransmission -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/transmission-4.0.3/third-party/libutp/include -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/transmission-4.0.3/third-party/jsonsl -I/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/transmission-4.0.3/third-party/wildmat -isystem /opt/local/include -isystem /opt/local/libexec/openssl3/include -isystem /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/transmission-4.0.3/cmake/../third-party/fast_float/include -isystem /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/build/third-party/libnatpmp.bld/pfx/include -isystem /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/build/third-party/dht.bld/pfx/include -isystem /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/transmission-4.0.3/cmake/../third-party/utfcpp/source -isystem /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/transmission-4.0.3/cmake/../third-party/wide-integer -isystem /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/transmission-4.0.3/cmake/../third-party/fmt/include -pipe -I/opt/local/libexec/openssl3/include -Os -DNDEBUG -I/opt/local/libexec/openssl3/include -isystem/opt/local/include -std=gnu++17 -arch ppc -mmacosx-version-min=10.6 -W -Wall -Wextra -Wcast-align -Wduplicated-cond -Wextra-semi -Wfloat-equal -Winit-self -Wint-in-bool-context -Wlogical-op -Wmissing-format-attribute -Wnull-dereference -Wpointer-arith -Wredundant-decls -Wredundant-move -Wrestrict -Wshadow -Wsign-compare -Wsuggest-override -Wuninitialized -Wunreachable-code -Wunused -Wunused-const-variable -Wunused-parameter -Wunused-result -Wwrite-strings -fobjc-arc -MD -MT libtransmission/CMakeFiles/transmission.dir/announce-list.cc.o -MF CMakeFiles/transmission.dir/announce-list.cc.o.d -o CMakeFiles/transmission.dir/announce-list.cc.o -c /opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/transmission-4.0.3/libtransmission/announce-list.cc
g++-mp-12: error: unrecognized command-line option '-fobjc-arc'; did you mean '-fobjc-gc'?
make[2]: *** [libtransmission/CMakeFiles/transmission.dir/announcer-udp.cc.o] Error 1
g++-mp-12: error: unrecognized command-line option '-fobjc-arc'; did you mean '-fobjc-gc'?
make[2]: *** [libtransmission/CMakeFiles/transmission.dir/announcer-http.cc.o] Error 1
g++-mp-12: error: unrecognized command-line option '-fobjc-arc'; did you mean '-fobjc-gc'?
make[2]: *** [libtransmission/CMakeFiles/transmission.dir/bitfield.cc.o] Error 1
g++-mp-12: error: unrecognized command-line option '-fobjc-arc'; did you mean '-fobjc-gc'?
make[2]: *** [libtransmission/CMakeFiles/transmission.dir/announce-list.cc.o] Error 1
g++-mp-12: error: unrecognized command-line option '-fobjc-arc'; did you mean '-fobjc-gc'?
make[2]: *** [libtransmission/CMakeFiles/transmission.dir/bandwidth.cc.o] Error 1
make[2]: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/build'
make[1]: *** [libtransmission/CMakeFiles/transmission.dir/all] Error 2
make[1]: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/build'
make: *** [all] Error 2
make: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/build'
Command failed:  cd "/opt/local/var/macports/build/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/work/build" && /usr/bin/make -j6 -w all VERBOSE=ON 
Exit code: 2
Error: Failed to build transmission-x11: command execution failed
Error: See /opt/local/var/macports/logs/_opt_PPCSnowLeopardPorts_net_transmission-x11/transmission-x11/main.log for details.
Error: Follow https://guide.macports.org/#project.tickets if you believe there
is a bug.
36-25% ission/CMakeFiles/transmission.dir/bitfield.cc.o] Error 1
zsh: no such file or directory: ission/CMakeFiles/transmission.dir/bitfield.cc.o]
g++-mp-12: error: unrecognized command-line option '-fobjc-arc'; did you mean '-fobjc-gc'?
make[2]: *** [libtransmission/CMakeFiles/transmission.dir/announce-list.cc.o] Error 1
g++-mp-12: error: unrecognized command-line option '-fobjc-arc'; did you mean '-fobjc-gc'?
make[2]: *** [libtransmission/CMakeFiles/transmission.dir/bandwidth.cc.o] Error 1
make[2]: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPor36-25% g++-mp-12: error: unrecognized command-line option '-fobjc-arc'; did you mean '-fobjc-gc'?
zsh: command not found: g++-mp-12:
zsh: no matches found: -fobjc-gc?
make[2]: Leaving directory `/opt/local/var/macports/build/_opt_PPCSnowLeopardPor36-25% make[2]: *** [libtransmission/CMakeFiles/transmission.dir/announce-list.cc.o] Error 1
```

### Which application of Transmission?

GTK+ app on Linux, BSD, etc.

### Which version of Transmission?

4.0.3

--- a/libtransmission/CMakeLists.txt
+++ b/libtransmission/CMakeLists.txt
@@ -10,11 +10,9 @@ endif()
 
 check_symbol_exists(SO_REUSEPORT "sys/types.h;sys/socket.h" HAVE_SO_REUSEPORT)
 
-check_cxx_compiler_flag(-fobjc-arc CXX_COMPILER_SUPPORTS_OBJC_ARC)
-
 add_compile_options(
 	# equivalent of XCODE_ATTRIBUTE_CLANG_ENABLE_OBJC_ARC YES for this directory
-	$<$<AND:$<BOOL:${APPLE}>,$<BOOL:${CXX_COMPILER_SUPPORTS_OBJC_ARC}>,$<COMPILE_LANGUAGE:C,CXX>>:-fobjc-arc>)
+	$<$<AND:$<BOOL:${APPLE}>,$<CXX_COMPILER_ID:AppleClang,Clang>,$<COMPILE_LANGUAGE:C,CXX>>:-fobjc-arc>)
 
 add_library(${TR_NAME} STATIC)
 
@@ -206,10 +204,10 @@ tr_allow_compile_if(
         watchdir-inotify.cc
     [=[[WITH_KQUEUE]]=]
         watchdir-kqueue.cc
-    [=[[APPLE AND CXX_COMPILER_SUPPORTS_OBJC_ARC]]=]
+    [=[[APPLE AND CMAKE_CXX_COMPILER_ID MATCHES "Clang"]]=]
         tr-assert.mm
         utils.mm
-    [=[[NOT (APPLE AND CXX_COMPILER_SUPPORTS_OBJC_ARC)]]=]
+    [=[[NOT (APPLE AND CMAKE_CXX_COMPILER_ID MATCHES "Clang")]]=]
         tr-assert.cc
     [=[[WIN32]]=]
         file-win32.cc

--- a/libtransmission/CMakeLists.txt
+++ b/libtransmission/CMakeLists.txt
@@ -10,9 +10,11 @@ endif()
 
 check_symbol_exists(SO_REUSEPORT "sys/types.h;sys/socket.h" HAVE_SO_REUSEPORT)
 
-add_compile_options(
-    # equivalent of XCODE_ATTRIBUTE_CLANG_ENABLE_OBJC_ARC YES for this directory
-    $<$<AND:$<BOOL:${APPLE}>,$<COMPILE_LANGUAGE:C,CXX>>:-fobjc-arc>)
+if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+    add_compile_options(
+        # equivalent of XCODE_ATTRIBUTE_CLANG_ENABLE_OBJC_ARC YES for this directory
+        $<$<AND:$<BOOL:${APPLE}>,$<COMPILE_LANGUAGE:C,CXX>>:-fobjc-arc>)
+endif()
 
 add_library(${TR_NAME} STATIC)
 
@@ -204,10 +206,10 @@ tr_allow_compile_if(
         watchdir-inotify.cc
     [=[[WITH_KQUEUE]]=]
         watchdir-kqueue.cc
-    [=[[APPLE]]=]
+    [=[[APPLE AND CMAKE_CXX_COMPILER_ID MATCHES "Clang"]]=]
         tr-assert.mm
         utils.mm
-    [=[[NOT APPLE]]=]
+    [=[[NOT (APPLE AND CMAKE_CXX_COMPILER_ID MATCHES "Clang")]]=]
         tr-assert.cc
     [=[[WIN32]]=]
         file-win32.cc

--- a/libtransmission/CMakeLists.txt
+++ b/libtransmission/CMakeLists.txt
@@ -11,8 +11,8 @@ endif()
 check_symbol_exists(SO_REUSEPORT "sys/types.h;sys/socket.h" HAVE_SO_REUSEPORT)
 
 add_compile_options(
-	# equivalent of XCODE_ATTRIBUTE_CLANG_ENABLE_OBJC_ARC YES for this directory
-	$<$<AND:$<BOOL:${APPLE}>,$<CXX_COMPILER_ID:AppleClang,Clang>,$<COMPILE_LANGUAGE:C,CXX>>:-fobjc-arc>)
+    # equivalent of XCODE_ATTRIBUTE_CLANG_ENABLE_OBJC_ARC YES for this directory
+    $<$<AND:$<BOOL:${APPLE}>,$<CXX_COMPILER_ID:AppleClang,Clang>,$<COMPILE_LANGUAGE:C,CXX>>:-fobjc-arc>)
 
 add_library(${TR_NAME} STATIC)
 

--- a/libtransmission/CMakeLists.txt
+++ b/libtransmission/CMakeLists.txt
@@ -10,11 +10,11 @@ endif()
 
 check_symbol_exists(SO_REUSEPORT "sys/types.h;sys/socket.h" HAVE_SO_REUSEPORT)
 
-if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
-    add_compile_options(
-        # equivalent of XCODE_ATTRIBUTE_CLANG_ENABLE_OBJC_ARC YES for this directory
-        $<$<AND:$<BOOL:${APPLE}>,$<COMPILE_LANGUAGE:C,CXX>>:-fobjc-arc>)
-endif()
+check_cxx_compiler_flag(-fobjc-arc CXX_COMPILER_SUPPORTS_OBJC_ARC)
+
+add_compile_options(
+	# equivalent of XCODE_ATTRIBUTE_CLANG_ENABLE_OBJC_ARC YES for this directory
+	$<$<AND:$<BOOL:${APPLE}>,$<BOOL:${CXX_COMPILER_SUPPORTS_OBJC_ARC}>,$<COMPILE_LANGUAGE:C,CXX>>:-fobjc-arc>)
 
 add_library(${TR_NAME} STATIC)
 
@@ -206,10 +206,10 @@ tr_allow_compile_if(
         watchdir-inotify.cc
     [=[[WITH_KQUEUE]]=]
         watchdir-kqueue.cc
-    [=[[APPLE AND CMAKE_CXX_COMPILER_ID MATCHES "Clang"]]=]
+    [=[[APPLE AND CXX_COMPILER_SUPPORTS_OBJC_ARC]]=]
         tr-assert.mm
         utils.mm
-    [=[[NOT (APPLE AND CMAKE_CXX_COMPILER_ID MATCHES "Clang")]]=]
+    [=[[NOT (APPLE AND CXX_COMPILER_SUPPORTS_OBJC_ARC)]]=]
         tr-assert.cc
     [=[[WIN32]]=]
         file-win32.cc

--- a/libtransmission/utils.cc
+++ b/libtransmission/utils.cc
@@ -274,7 +274,7 @@ double tr_getRatio(uint64_t numerator, uint64_t denominator)
 
 // ---
 
-#ifndef __APPLE__
+#if !(defined(__APPLE__) && defined(__clang__))
 
 std::string tr_strv_convert_utf8(std::string_view sv)
 {

